<a href="https://colab.research.google.com/github/divyansh1195/Halliburton-Landmark-Learning/blob/master/Data%20Wrangling%20on%20Oil%20and%20Gas%20dataset%20using%20Python/Tidying_and_Transformation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### In this notebook, I have performed Data Tidying and Transformation on Oil and Gas dataset (Source:Halliburton Landmark Learning). The notebook implementation has been done based on tutorials provided in the Landmark Learning course, "Data Science using R." 

#### I had completed the course, "Data Science using R" from Landmark Learning and then tried to implement the data wrangling steps on my own in Python. 

# **Tidying and Transformation**

## Tidying+Transforming=Data Wrangling

## Convert dataset from a wider format to a longer format

In [1]:
import pandas as pd
#read excel file
untidy_data=pd.read_excel('Production_untidy_data.xlsx')

In [2]:
untidy_data.head()

State  2004  2005  2006    2007    2008    2009
0     Alabama    25    25    28    23.0    17.0    26.0
1      Alaska    11    15    17    16.0     8.0    10.0
2    Arkansas   246   247   251   271.0   256.0   276.0
3     Arizona     4     1     2     5.0     2.0     2.0
4  California  3897  4174  4389  4776.0  5291.0  5544.0

In [3]:
untidy_data.shape

(33, 7)

In [4]:
tidy_data = pd.melt(untidy_data, id_vars=['State'], 
                  var_name='Year', value_name='rate_class_1_oil')
tidy_data.head()

State  Year  rate_class_1_oil
0     Alabama  2004              25.0
1      Alaska  2004              11.0
2    Arkansas  2004             246.0
3     Arizona  2004               4.0
4  California  2004            3897.0

In [5]:
tidy_data.shape

(198, 3)

So, the untidy dataset has been transformed from (33,7) to (198,3)

## Now, I want to convert the dataset from longer format to wider format

In [6]:
wider_data = tidy_data.pivot_table('rate_class_1_oil', ['State'], 'Year')
wider_data.head()
#here, the transformed dataset has been converted back to it's original format

Year          2004    2005    2006    2007    2008    2009
State                                                     
Alabama       25.0    25.0    28.0    23.0    17.0    26.0
Alaska        11.0    15.0    17.0    16.0     8.0    10.0
Arizona        4.0     1.0     2.0     5.0     2.0     2.0
Arkansas     246.0   247.0   251.0   271.0   256.0   276.0
California  3897.0  4174.0  4389.0  4776.0  5291.0  5544.0

# Organizing the Dataset

In [7]:
untidy_1_2=pd.read_excel('Production_untidy_data_reshape_rate_class_1,2_oil.xlsx')
print(untidy_1_2.shape)
print(untidy_1_2.isnull().sum())

(66, 8)
State    0
Class    0
2004     0
2005     0
2006     0
2007     2
2008     2
2009     3
dtype: int64


In [8]:
untidy_1_2=untidy_1_2.fillna(0)#fiil the missing values as 0
print(untidy_1_2.isnull().sum())
print(untidy_1_2.shape)

State    0
Class    0
2004     0
2005     0
2006     0
2007     0
2008     0
2009     0
dtype: int64
(66, 8)


In [9]:
untidy_1_2.dtypes

State     object
Class      int64
2004       int64
2005       int64
2006       int64
2007     float64
2008     float64
2009     float64
dtype: object

In [10]:
df = pd.melt(untidy_1_2, id_vars=['State', 'Class'], 
                  var_name='Year', value_name='rate_class_oil_value')
print(df.dtypes)
print(df.head())

State                    object
Class                     int64
Year                     object
rate_class_oil_value    float64
dtype: object
        State  Class  Year  rate_class_oil_value
0     Alabama      1  2004                  25.0
1      Alaska      1  2004                  11.0
2    Arkansas      1  2004                 246.0
3     Arizona      1  2004                   4.0
4  California      1  2004                3897.0


In [11]:
df['Year']=df['Year'].astype('int64')#convert Year from object to int64
df['rate_class_oil_value']=df['rate_class_oil_value'].astype('int64')#convert 'rate_class_oil_value' from float64 to int64
print(df.dtypes)

State                   object
Class                    int64
Year                     int64
rate_class_oil_value     int64
dtype: object


In [12]:
df.isnull().sum()#again check that there are no missing values in the dataset columns

State                   0
Class                   0
Year                    0
rate_class_oil_value    0
dtype: int64

In [13]:
df.head()

State  Class  Year  rate_class_oil_value
0     Alabama      1  2004                    25
1      Alaska      1  2004                    11
2    Arkansas      1  2004                   246
3     Arizona      1  2004                     4
4  California      1  2004                  3897

## Let's answer some questions related to the dataset, 'df'

In [14]:
## Total no. of wells per state wrt each year and each class
X = df.groupby(['State', 'Year','Class'])['rate_class_oil_value'].sum().to_frame()
X = X.pivot_table('rate_class_oil_value', ['State'], ['Year','Class'])
print(X.shape)
X

(33, 12)


Year            2004          2005         ...   2008          2009       
Class              1      2      1      2  ...      1      2      1      2
State                                      ...                            
Alabama           25     12     25     12  ...     17     22     26     19
Alaska            11      4     15      5  ...      8      3     10      6
Arizona            4      1      1      1  ...      2      2      2      0
Arkansas         246    206    247    208  ...    256    252    276    265
California      3897   3662   4174   4118  ...   5291   4806   5544   4936
Colorado         701    703    733    699  ...    885    710    937    717
FG                46     25     67     38  ...     41     14     46     23
FP                 0      2      2      2  ...      5      3      3      0
Florida            3      0      2      1  ...      1      1     20     11
Kansas         22558   7194  23415   7202  ...  24275   7806  24654   7943
Kentucky        3083    704   3513    576  ...   3243    675      0      0
Louisiana      11430   1389  11604   1441  ...  11870   1552  11987   1601
Maryland           0      0      0      0  ...      0      0      0      0
Michigan        1223    670   1334    621  ...   1226    695   1198    655
Mississippi      109     48    107     54  ...    129     66    142     83
Missouri         143      5    188     23  ...    239     64    287     52
Montana         1061    448   1090    474  ...   1237    474   1253    470
ND               148    114    134    119  ...    127    147    180    151
Nebraska          50     35     41     36  ...     50     29     53     36
Nevada             7      4      7      5  ...      7      4      5      4
New Mexico      3091   1767   3005   1837  ...   3294   2217   3518   2136
New York        2878     43   2605     81  ...   2773    388   2547    383
Ohio            7857   1625   7676   1583  ...   8155   1583   8373   1658
Oklahoma       17910   5032  17987   5181  ...  18158   5212  17650   5227
Oregon             0      0      0      0  ...      0      0      0      0
Pennsylvania    5062    535   5251    347  ...   4060    639   4053    588
South Dakota       1      2      1      6  ...      4      6      3      5
Tennessee        420    129    411    130  ...      0      0      0      0
Texas          42193  16144  41549  16187  ...  42401  17219  44021  17679
Utah             228     63    221     79  ...    177    102    206     80
Virginia           3      1      5      2  ...      4      1      3      1
West Virginia   3129    373   2655    421  ...   3396    452   2597    409
Wyoming         2184    945   2275    989  ...   2309   1038   2193   1099

[33 rows x 12 columns]

In [15]:
##Total no. of wells per state wrt each year i.e. combine wells in class 1 and 2 for each state
X=df.groupby(['State', 'Year'])['rate_class_oil_value'].sum().to_frame()
X = X.pivot_table('rate_class_oil_value', ['State'], 'Year')
print(X.shape)
X

(33, 6)


Year            2004   2005   2006   2007   2008   2009
State                                                  
Alabama           37     37     51     46     39     45
Alaska            15     20     18     18     11     16
Arizona            5      2      3      7      4      2
Arkansas         452    455    486    486    508    541
California      7559   8292   8610   9222  10097  10480
Colorado        1404   1432   1462   1462   1595   1654
FG                71    105     76     70     55     69
FP                 2      4      4      2      8      3
Florida            3      3      3      1      2     31
Kansas         29752  30617  31235  31465  32081  32597
Kentucky        3787   4089   4155   4306   3918      0
Louisiana      12819  13045  13009  13084  13422  13588
Maryland           0      0      0      0      0      0
Michigan        1893   1955   1963   1912   1921   1853
Mississippi      157    161    169    190    195    225
Missouri         148    211    202    213    303    339
Montana         1509   1564   1600   1613   1711   1723
ND               262    253    255    271    274    331
Nebraska          85     77     72     77     79     89
Nevada            11     12      8      7     11      9
New Mexico      4858   4842   5087   5261   5511   5654
New York        2921   2686   2889   2650   3161   2930
Ohio            9482   9259   9497   8550   9738  10031
Oklahoma       22942  23168  23315  23287  23370  22877
Oregon             0      0      0      0      0      0
Pennsylvania    5597   5598   6170   4795   4699   4641
South Dakota       3      7      8      7     10      8
Tennessee        549    541    127      0      0      0
Texas          58337  57736  57765  58457  59620  61700
Utah             291    300    313    274    279    286
Virginia           4      7      8      3      5      4
West Virginia   3502   3076   3340   3520   3848   3006
Wyoming         3129   3264   3277   3396   3347   3292

This shows that for Alabama in 2004 there were total 37 wells.

25 'Class 1' wells and 12 'Class 2' wells

In [16]:
## Total number of wells per state wrt each class
X=df.groupby(['State', 'Class'])['rate_class_oil_value'].sum().to_frame()
X = X.pivot_table('rate_class_oil_value', ['State'], 'Class')
print(X.shape)
X

(33, 2)


Class               1       2
State                        
Alabama           144     111
Alaska             77      21
Arizona            16       7
Arkansas         1547    1381
California      28071   26189
Colorado         4883    4126
FG                306     140
FP                 13      10
Florida            29      14
Kansas         142497   45250
Kentucky        17221    3034
Louisiana       70070    8897
Maryland            0       0
Michigan         7493    4004
Mississippi       730     367
Missouri         1189     227
Montana          6899    2821
ND                848     798
Nebraska          285     194
Nevada             36      22
New Mexico      19292   11921
New York        16176    1061
Ohio            47030    9527
Oklahoma       107523   31436
Oregon              0       0
Pennsylvania    28413    3087
South Dakota       13      30
Tennessee         900     317
Texas          253253  100362
Utah             1231     512
Virginia           23       8
West Virginia   17913    2379
Wyoming         13619    6086

This shows that there are 144 'Class 1' wells in Alabama and 111 'Class 2' wells